dll/win32/shell32/CSendToMenu.cpp

```diff
@@ -0,0 +1,520 @@
+
+    BOOL bShift = (GetAsyncKeyState(VK_SHIFT) < 0);
+    BOOL bCtrl = (GetAsyncKeyState(VK_CONTROL) < 0);
+
+    // THIS CODE IS NOT HUMAN-FRIENDLY. SORRY.
+    // (We have to translate a SendTo action to a Drop action)
+    DWORD dwKeyState = MK_LBUTTON;
+    if (bShift && bCtrl)
+        dwKeyState |= MK_SHIFT | MK_CONTROL;
+    else if (!bShift)
+        dwKeyState |= MK_CONTROL;
+    if (bCtrl)
+        dwKeyState |= MK_SHIFT;
+
+    POINTL ptl = { 0, 0 };
+    HRESULT hr = pDropTarget->DragEnter(pDataObject, dwKeyState, ptl, &dwEffect);
+    if (SUCCEEDED(hr) && dwEffect != DROPEFFECT_NONE)
+    {
+        // THIS CODE IS NOT HUMAN-FRIENDLY. SORRY.
+        // (We have to translate a SendTo action to a Drop action)
+        if (bShift && bCtrl)
+            dwEffect = DROPEFFECT_LINK;
+        else if (!bShift)
+            dwEffect = DROPEFFECT_MOVE;
+        else
+            dwEffect = DROPEFFECT_COPY;
+
+        hr = pDropTarget->Drop(pDataObject, dwKeyState, ptl, &dwEffect);
+    }
+    else
+    {
+        ERR("DragEnter: %08lX\n", hr);
+        pDropTarget->DragLeave();
+    }
+
+    return hr;
+}
+
+// get an IShellFolder from CSIDL
+IShellFolder *
+CSendToMenu::GetSpecialFolder(HWND hwnd, int csidl, LPITEMIDLIST *ppidl)
+{
+    if (ppidl)
+        *ppidl = NULL;
+
+    if (!m_pDesktop)
+    {
+        SHGetDesktopFolder(&m_pDesktop);
+        if (!m_pDesktop)
+        {
+            ERR("SHGetDesktopFolder\n");
+            return NULL;
+        }
+    }
+
+    LPITEMIDLIST pidl = NULL;
+    HRESULT hr = SHGetSpecialFolderLocation(hwnd, csidl, &pidl);
+    if (FAILED(hr))
+    {
+        ERR("SHGetSpecialFolderLocation: %08lX\n", hr);
+        return NULL;
+    }
+
+    IShellFolder *pFolder = NULL;
+    hr = m_pDesktop->BindToObject(pidl, NULL, IID_PPV_ARG(IShellFolder, &pFolder));
+
+    if (ppidl)
+        *ppidl = pidl;
+    else
+        CoTaskMemFree(pidl);
+
+    if (SUCCEEDED(hr))
+        return pFolder;
+
+    ERR("BindToObject: %08lX\n", hr);
+    return NULL;
+}
+
+// get a UI object from PIDL
+HRESULT CSendToMenu::GetUIObjectFromPidl(HWND hwnd, LPITEMIDLIST pidl,
+                                         REFIID riid, LPVOID *ppvOut)
+{
+    *ppvOut = NULL;
+
+    LPCITEMIDLIST pidlLast;
+    CComPtr<IShellFolder> pFolder;
+    HRESULT hr = SHBindToParent(pidl, IID_PPV_ARG(IShellFolder, &pFolder), &pidlLast);
+    if (FAILED(hr))
+    {
+        ERR("SHBindToParent: %08lX\n", hr);
+        return hr;
+    }
+
+    hr = pFolder->GetUIObjectOf(hwnd, 1, &pidlLast, riid, NULL, ppvOut);
+    if (FAILED(hr))
+    {
+        ERR("GetUIObjectOf: %08lX\n", hr);
+    }
+    return hr;
+}
+
+void CSendToMenu::UnloadItem(SENDTO_ITEM *pItem)
+{
+    if (!pItem)
+        return;
+
+    CoTaskMemFree(pItem->pidlChild);
+    CoTaskMemFree(pItem->pszText);
+    DestroyIcon(pItem->hIcon);
+    HeapFree(GetProcessHeap(), 0, pItem);
+}
+
+void CSendToMenu::UnloadAllItems()
+{
+    SENDTO_ITEM *pItems = m_pItems;
```

Comment:
Why is SENDTO_ITEM implemented as a linked list?
This is c++ code, not c?

Reply:
I don't wanna write copy constructor for it.